### PR TITLE
8263583: Emoji rendering on macOS

### PR DESCRIPTION
**Pull request by JB-Dmitry**
This is the implementation used by JetBrains Runtime for the last 4 years, after some cleanup, and with one problem,
found while preparing the pull request, fixed.
Even though typical scenarios for a UI application should be covered, it's not a complete solution. In particular, emoji-s
still won't be rendered for large font sizes (more than 100pt), and for non-trivial composite/painting modes.
Notable implementation details are listed below.

**Glyph image generation**

Deprecated CGContextShowGlyphsAtPoint function, used by JDK on macOS to render text, cannot render emojis,
CTFontDrawGlyphs is used instead. It ignores the scale component of text transformation matrix, so a 'real-sized'
CTFont object should be passed to it. The same sizing procedure is done when calculating glyph metrics, because they
are not scaled proportionally with font size (as they do for vector fonts).

**Glyph image storage**

Existing GlyphInfo structure is used to store color glyph image. Color glyph can be distinguished by having 4 bytes
of storage per pixel. Color components are stored in pre-multiplied alpha format.

**Glyph rendering**

Previously, GlyphList instance always contained glyphs in the same format (solid, grayscale or LCD), determined by the
effective rendering hint. Now the renderers must be prepared to GlyphList having 'normal' glyphs interspersed with
color glyphs (they can appear due to font fallback). This isn't a problem for OpenGL renderer (used for on-screen painting),
but GlyphListLoopPipe-based renderers (used for off-screen painting) needed an adjustment to be able to operate on
specific segments of GlyphList.
As an incidental optimization, calculation of GlyphList bounds ('getBounds' method) is performed now only when needed
(most text renderers don't need this information).
Speaking of the actual rendering of the glyph image, it's done by the straightforward glDrawPixels call in OpenGL renderer,
and by re-using existing Blit primitive in off-screen renderers.

**Testing**

There's no good way to test the new functionality automatically, but I've added a test verifying that 'something' is
rendered for the emoji character, when painting to BufferedImage.

Existing tests pass after the change.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8263583](https://bugs.openjdk.java.net/browse/JDK-8263583): Emoji rendering on macOS

### Reviewers
 * [Sergey Bylokhov](https://openjdk.java.net/census#serb) (@mrserb - **Reviewer**)
 * [Phil Race](https://openjdk.java.net/census#prr) (@prrace - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3007/head:pull/3007` \
`$ git checkout pull/3007`

Update a local copy of the PR: \
`$ git checkout pull/3007` \
`$ git pull https://git.openjdk.java.net/jdk pull/3007/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3007`

View PR using the GUI difftool: \
`$ git pr show -t 3007`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3007.diff">https://git.openjdk.java.net/jdk/pull/3007.diff</a>

</details>

### CODE REVIEW

**Comment by avu** [comment]
Looks good

**Comment by prrace** [comment]
Initial testing it looks OK with OpenGL but it renders as garbage with the metal pipeline.
I'd guess metal isn't expecting 4bpp and the stride is wrong.
Since metal is now integrated this should be fixed before this change can be pushed.

File: src/java.desktop/share/classes/sun/java2d/loops/DrawGlyphListColor.java [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    public void DrawGlyphListColor(SunGraphics2D sg2d, SurfaceData dest,
+                                   GlyphList srcData,
+                                   int fromGlyph, int toGlyph) {
+        // actual implementation is in the 'General' subclass
+    }
+
+    // This instance is used only for lookup.
+    static {
+        GraphicsPrimitiveMgr.registerGeneral(
+                                new DrawGlyphListColor(null, null, null));
```
Comment (mrserb) [inline src/java.desktop/share/classes/sun/java2d/loops/DrawGlyphListColor.java:71]: If it is a complete "General" loop which supports all possible combination of incoming src/comp/dst/transform/etc I suggest to confirm that in the test.

**Comment by JB-Dmitry** [comment]
> Initial testing it looks OK with OpenGL but it renders as garbage with the metal pipeline.
> I'd guess metal isn't expecting 4bpp and the stride is wrong.
> I see updates in OGLTextRenderer so something similar likely needed for metal.
> Since metal is now integrated this should be fixed before this change can be pushed.

I'm going to add Metal implementation soon.

**Comment by JB-Dmitry** [comment]
> If it is a complete "General" loop which supports all possible combination of incoming src/comp/dst/transform/etc I suggest to confirm that in the test.

@mrserb This code is currently only used by GlyphListLoopPipe-based text renderers (SolidTextRenderer, AATextRenderer, LCDTextRenderer), so it won't be invoked for arbitrary composites and paint types. But different destination types (image formats) and transforms are supported.

What kind of test are you having in mind? I can extend the added MacEmoji.java test, that will check that 'something' is painted to different types of BufferedImage-s and/or with transforms applied. Or you want me to add a manual test case?

**Comment by mrserb** [comment]
> What kind of test are you having in mind? I can extend the added MacEmoji.java test, that will check that 'something' is painted to different types of BufferedImage-s and/or with transforms applied. Or you want me to add a manual test case?

The automated tests will be enough, not sure about 'something is painted' however, it is better to check that it works.

**Comment by JB-Dmitry** [comment]
@mrserb I don't know how to check automatically that glyph painting works correctly. Could you please suggest a way to do it? In JetBrains Runtime we have a test that checks that rendered emoji glyph matches one of stored 'golden images', but I don't think that's suitable for OpenJDK, unless someone will volunteer to update that list when the need arises (e.g. when newer version of macOS changes the font, or rendering details). At the moment, we have already 7 golden images for a single glyph in our repository.

**Comment by mrserb** [comment]
You the current image from the MacEmoji test as a golden image for other formats/transforms/extraalpha/etc. For example, if DST type is changed then it is unlikely the shape of the emoji will be changed as well, or if it is too big/ too small.

BTW It will be good if the test will fail on the current metal implementation.

**Comment by JB-Dmitry** [comment]
Implementation for Metal pipeline has been added.

I've also updated the test case to check rendering into different types of images, including VolatileImage. The test will fail now, if corresponding OpenGL (or Metal, if it's explicitly enabled) implementation part would be rolled back.

**Comment by avu** [comment]
Looks good!

**Comment by mrserb** [comment]
It looks fine to me. I will run the tests in CI just in case.

**Review by mrserb — APPROVED** [review]

**Comment by JB-Dmitry** [comment]
Any chance for a second review?

**Review by prrace — APPROVED** [review]

File: src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLTextRenderer.m [hunk 2]
```diff
@@ -324,6 +324,19 @@ void MTLTR_FreeGlyphCaches() {
     }
 }
 
+MTLPaint* storedPaint = nil;
```
[the remark this answers is not in the document]
Comment (JB-Dmitry) [reply src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLTextRenderer.m]: Added 'static', thanks.

**Review by prrace — APPROVED** [review]

**Comment by JB-Dmitry** [comment]
@mrserb Could you please re-approve the request? There's been only a tiny change since you've approved it earlier.

Do I understand correctly that this re-approval is required to integrate the change? Github shows the request as having two approvals, and the 'ready' tag. Or is it just openjdk bot not implementing the workflow correctly?

**Review by mrserb — APPROVED** [review]

**Comment by mrserb** [comment]
It is not necessary to get reapprove for push, it is just a good style to do.

**Pull request closed without merging**